### PR TITLE
[SPARK-11090] [SQL] Constructor for Product types from InternalRow

### PR DESCRIPTION
**Pull request by marmbrus**
This is a first draft of the ability to construct expressions that will take a catalyst internal row and construct a Product (case class or tuple) that has fields with the correct names.  Support include:
- Nested classes
- Maps
- Efficiently handling of arrays of primitive types

Not yet supported:
- Case classes that require custom collection types (i.e. List instead of Seq).

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/Encoder.scala [hunk 1]
```diff
@@ -17,6 +17,8 @@
 
 package org.apache.spark.sql.catalyst.encoders
 
+import org.apache.spark.sql.catalyst.expressions.{Attribute, AttributeReference}
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/Encoder.scala]: nit order

**Comment by rxin** [comment]
High level approach LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala [hunk 2]
```diff
@@ -114,6 +118,298 @@ trait ScalaReflection {
...
+          case t if t <:< definitions.LongTpe => Some(classOf[java.lang.Long])
+          case t if t <:< definitions.DoubleTpe => Some(classOf[java.lang.Double])
+          case t if t <:< definitions.FloatTpe => Some(classOf[java.lang.Float])
+          case t if t <:< definitions.ShortTpe => Some(classOf[java.lang.Short])
+          case t if t <:< definitions.ByteTpe => Some(classOf[java.lang.Byte])
+          case t if t <:< definitions.BooleanTpe => Some(classOf[java.lang.Boolean])
+          case _ => None
+        }
+
+        boxedType.map { boxedType =>
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala]: Can we use different names for these two?

**Comment by marmbrus** [comment]
Going to merge and address any comments in a follow-up.

**Pull request closed without merging**